cli/cmd/healthcheck_extensions.go

```diff
@@ -0,0 +1,315 @@
+package cmd
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/briandowns/spinner"
+	"github.com/linkerd/linkerd2/pkg/healthcheck"
+	"github.com/linkerd/linkerd2/pkg/version"
+	"github.com/mattn/go-isatty"
+	utilsexec "k8s.io/utils/exec"
+)
+
+// glob is satisfied by filepath.Glob.
+type glob func(string) ([]string, error)
+
+// extension contains the full path of an extension executable. If it's a
+// a built-in extension, the builtin field is populated with its name.
+type extension struct {
+	path    string
+	builtin string
+}
+
+var (
+	builtInChecks = map[string]struct{}{
+		"jaeger":       {},
+		"multicluster": {},
+		"viz":          {},
+	}
+)
+
+// findExtensions searches the path for all linkerd-* executables and returns a
+// slice of check commands, and a slice of missing checks.
+func findExtensions(pathEnv string, glob glob, exec utilsexec.Interface, nsLabels []string) ([]extension, []string) {
+	cliExtensions := findCLIExtensionsOnPath(pathEnv, glob, exec)
+
+	// first, collect config extensions that are "always" enabled
+	extensions, checksSeen := findAlwaysChecks(cliExtensions, exec)
+
+	labelMap := map[string]struct{}{}
+	for _, label := range nsLabels {
+		if _, ok := checksSeen[label]; !ok {
+			labelMap[label] = struct{}{}
+		}
+	}
+
+	// second, collect on-cluster extensions
+	for _, e := range cliExtensions {
+		suffix := suffix(e)
+		if _, ok := labelMap[suffix]; ok {
+			extensions = append(extensions, extension{path: e})
```

Comment:
does this double-count cliExtensions which are "Always" and also installed on the cluster?

Reply:
that's handled in `findExtensions`, anything returned by `checksSeen` is not put into the `nsLabelSet`, so at that point we're done searching for an executable for that namespace. i'll add a comment where we declare `nsLabelSet` to clarify.